torch/autograd/__init__.py

```diff
@@ -26,24 +26,63 @@
 __all__ = ['Variable', 'Function', 'backward', 'grad_mode']
 
 _OptionalTensor = Optional[torch.Tensor]
+_NtSize = Union[Tuple[torch.Tensor], List[torch.Tensor]]
+_ShapeorNestedShape = Union[_NtSize, _size]
+
+
+
+# Lower this to c++?
+def equal_nested_tensor_sizes(nt_size_1: Tuple, nt_size_2: Tuple) -> bool:
+    for chunk_1, chunk_2 in zip(nt_size_1, nt_size_2):
+        if not torch.equal(chunk_1, chunk_2):
+            return False
+    return True
+
+
+def equal_shapes(size_1: _ShapeorNestedShape, size_2: _ShapeorNestedShape) -> bool:
+    if isinstance(size_1, tuple) and isinstance(size_2, tuple):
+        return equal_nested_tensor_sizes(size_1, size_2)
+    return size_1 == size_2
+
+
+def _calculate_shape(output: torch.Tensor, grad: torch.Tensor,
+                     is_grads_batched: bool) -> Tuple[_ShapeorNestedShape, _ShapeorNestedShape]:
+    if output.is_nested:
+        if is_grads_batched:
+            raise RuntimeError("Batched grads are not supported with Nested Tensor.")
+        grad_shape = grad.nested_size()
+        out_shape = output.nested_size()
+        return out_shape, grad_shape
+
+    reg_out_shape = output.shape
+    reg_grad_shape = grad.shape if not is_grads_batched else grad.shape[1:]
+    return reg_out_shape, reg_grad_shape
 
 def _make_grads(outputs: Sequence[torch.Tensor], grads: Sequence[_OptionalTensor],
                 is_grads_batched: bool) -> Tuple[_OptionalTensor, ...]:
     new_grads: List[_OptionalTensor] = []
     for out, grad in zip(outputs, grads):
         if isinstance(grad, torch.Tensor):
-            grad_shape = grad.shape if not is_grads_batched else grad.shape[1:]
-            if not out.shape == grad_shape:
+            if out.is_nested != grad.is_nested:
```

Comment:
I don't think we should special case for nested Tensor here.
If you need to check shapes in a way that should be overrideable, you should create a new native function `has_same_shape(t1, t2)` that you can then override for NestedTensor.
You can then use this function when the input is not a plain Tensor (subclass like).